js/logic.js

```diff
@@ -0,0 +1,862 @@
+                        new Error().stack);
+      }
+    }
+    return scope;
+  }
+
+  /**
+   * Most often, scopes and namespaces map one-to-one with class instances. With
+   * defineScope(), you can associate a Scope with an object, and then use that
+   * object in place of the scope. For instance:
+   *
+   *   function MyClass() {
+   *     logic.defineScope(this, 'MyClass');
+   *     logic.event(this, 'initialized');
+   *   }
+   */
+  logic.defineScope = function(obj, namespace, defaultDetails) {
+    // Default to the object's class name, if available.
+    if (!namespace && obj && obj.constructor && obj.constructor.name) {
+      namespace = obj.constructor.name;
+    }
+    var scope = new Scope(namespace, defaultDetails);
+    objectToScope.set(obj, scope);
+    return scope;
+  };
+
+  /**
+   * Sometimes, you may want to log several events, each with shared
+   * details. With logic.subscope(), you can create a child scope that
+   * shares the same namespace, but adds additional default details
+   * onto each message. For instance:
+   *
+   *   logic.defineScope(this, 'Account', { accountId: 1 });
+   *   var scope = logic.subscope(this, { action: 'move' });
+   *   logic.log(scope, 'start');
+   *   // event: Account/start { accountId: 1, action: 'move' }
+   */
+  logic.subscope = function(scope, defaultDetails) {
+    scope = toScope(scope);
+    return new Scope(scope.namespace, into(clone(scope.defaultDetails),
+                                           clone(defaultDetails)));
+  };
+
+  /**
+   * Emit an event. `logic(...)` is shorthand for `logic.event(...)`.
+   * See the module docs for more about events.
+   *
+   * @param {Scope} scope
+   *   The scope (i.e. "namespace") for this event.
+   * @param {string} type
+   *   A string, typically camelCased, describing the event taking place.
+   * @param {object} details
+   *   Optional details about this event, such as identifiers or parameters.
+   *   These details will be mixed in with any default details specified
+   *   by the Scope.
+   */
+  logic.event = function(scope, type, details) {
+    scope = toScope(scope);
+
+    if (typeof type !== 'string') {
+      throw new Error('Invalid "type" passed to logic.event(); ' +
+                      'expected a string, got "' + type + '"');
+    }
+
+    if (scope.defaultDetails) {
+      if(isPlainObject(details)) {
+        details = into(clone(scope.defaultDetails), clone(details));
+      } else {
+        details = clone(scope.defaultDetails);
+      }
+    } else {
+      details = clone(details);
+    }
+
+    var event = new LogicEvent(scope, type, details);
+
+    emitter.emit(event.type, event);
+    emitter.emit('*', event);
+
+    if (realtimeLogEverything) {
+      dump('logic: ' + event.toString() + '\n');
+    }
+
+    return event;
+  };
+
+  var nextId = 1;
+
+  /**
+   * Return a sequential unique-for-this-process identifier.
```

Comment:
Nit: "unique-for-this-process" is a fancier guarantee than what's currently implemented.  Since we know this code already runs in a situation where it's not true (main page, worker), this should probably be weakened to "unique-for-all-users-of-this-module-instance".